### ISSUE TITLE
Metadata 

### ISSUE THREAD

**Issue by jsulinski** [JCBarry/argyle#4]
Is there any way to access the metadata.account_id parameter in the callback controller?

**Comment by JCBarry** [comment]
Can you give me an example of what you're looking for?  I searched the Plaid docs for the `metadata.account_id` field and wasn't clear what you're after.  I'm sure there is a way to access the data you're looking for in that controller.

Thanks!

**Comment by jsulinski** [comment]
It seems that to actually get a bank account token back you need to pass
public key and account_id.

As far as I know, the only way to get that is through the callback from the
plaid js, it is passed as metadata.account_id.

Thanks for any help!

James

**Comment by jsulinski** [comment]
To be more clear, I'm wondering if you have successfully received a stripe_bank_account_token from Plaid? If so, can you perhaps share an example of the code used to do so?

**Comment by JCBarry** [comment]
Hey! Alright, you're using Stripe -- I haven't used Argyle for this purpose -- but I can't see why it wouldn't work.  You can try to take a look at the `params` in the controller.  I suspect you should see the account ID in addition to the public token.

I'll take a look at this as well.

**Comment by jsulinski** [comment]
Okay! That makes sense. I ended up working around Argyle, replacing the plaid_link with js/html from Plaid's docs, this allows me to customize the onSuccess function and send the account_id as :account_id. 

The challenge I'm facing now, is that when I use the client Argyle passes to exchange the token, with that access_id plaid just passed (required to get a bank account token), Plaid says 'account not found'. 

I have one idea -- perhaps creating the 'user/account' on Plaid first is necessary, before exchanging the token.

**Comment by JCBarry** [comment]
Doing a quick test, I was able to see that "account_id" will be available your callback controller as `params[:account_id]`.

Here is a look at the raw request params to the callback using Argyle:

```
{
"utf8"=>"✓", 
"authenticity_token"=>"Si5StRJt234VEDnxihQhcGuKpqOJ/QFYM2342ufOyk=", 
"account_id"=>"", 
"institution"=>{"name"=>"Chase", "type"=>"chase"}, 
"public_token"=>"test,chase,connected", 
"controller"=>"user/settings/plaid", 
"action"=>"callback"
}
```

I believe it is empty because I don't have Stripe enabled on our account.

In order to use this, we should update the gem and change the `plaid_exchange_token` method to optionally send the `account_id`.  In the meantime, you could override the method in your controller to make the request like this:

``` ruby
def plaid_exchange_token
  @plaid_access_token_response ||= Argyle.plaid_client.exchange_token(plaid_public_token, account_id: params[:account_id])
end
```

**Comment by jsulinski** [comment]
Thanks so much for the help. That is precisely what I did. The problem is that, when I try to use Argyle's plaid_client to do the exchange, I get the error:

[2016-05-11T13:06:58.301870 #17658]  INFO -- : Completed 500 Internal Server Error in 494ms (ActiveRecord: 10.0ms)
F, [2016-05-11T13:06:58.304443 #17658] FATAL -- : 
Plaid::NotFound (account not found):
  app/controllers/plaid_controller.rb:51:in `plaid_exchange_token'

Whereas when I simply use a Faraday request to exchange the token myself, it actually works, passing exactly the same data. 

Any ideas?

**Comment by JCBarry** [comment]
When I last published the gem, `plaid-ruby` was stable at 1.7.1.  It looks like the `account_id` options were added in the 2.0.0.alpha version of the gem.

It also looks like they made _massive_ changes to the gem -- so this will need some testing anyway.

**Comment by jsulinski** [comment]
Thanks so much for this!

Just to be clear, when you checked earlier for the account_id being passed along, was that with the alpha version of the gem? Because I was pretty sure I tried using params[:account_id] and it wasn't accessible, using 1.7.x. 

James

**Comment by jsulinski** [comment]
It does look like there are a lot of breaking changes in 2.x, including parameter changes for 3+ functions.

**Comment by JCBarry** [comment]
Yeah lots of breaking changes in the 2.0 release of the gem -- I'll start working through those.  It's still in alpha but it looks like they are prepping for a stable release shortly.

The `account_id` being passed along isn't contingent on the gem version -- when the user finishes going through the Plaid flow modal, it will post to the form.  The JS loaded directly from Plaid takes care of this and I would assume it added this field to the form when they made the Stripe announcement.

**Comment by JCBarry** [comment]
Check out the 2.0 branch in your project and see if it solves some issues for you -- tests are passing and following the `plaid-ruby` upgrade document I think I covered everything.

If you wouldn't mind giving it a shot that would be helpful.

**Comment by JCBarry** [comment]
I released a new version of the gem `1.0.0.beta` that includes the recently released 2.0 version of the official plaid ruby gem which has support for the `account_id` user exchange.

Unfortunately, I don't have access to Stripe ACH with Plaid so I can't fully test that flow.
